### ISSUE TITLE
Looking for an internal design document

### ISSUE THREAD

**Issue by sqsw** [vespa-engine/vespa#5434]
I'm looking into vespa for a big data processing and serving senario, it might look good to fit into our use case. Before deep dive the massive source code to make an evaluation, could you please share some piece of internal architecture design document for me to get a better understanding? I failed to find it on http://docs.vespa.ai/, please point me to the right place if you already published the internal design. Thank you in advance.

**Comment by sqsw** [comment]
@bratseth

**Comment by sqsw** [comment]
Specifically,  I'd be interested to know the following issues,
1.  How to create the index when feeding document, and how to store the indexes in RAM or disk? 
2. How to lookup the index and find the match for a search index query?

**Comment by kkraune** [comment]
I think http://docs.vespa.ai/documentation/elastic-vespa.html and http://docs.vespa.ai/documentation/querying-vespa.html are relevant docs to understand feeding and querying

**Comment by bratseth** [comment]
Yes, we don't have detailed and updated design docs, and you don't need them to use Vespa. To evaluate I suggest you just try to create your application and do some benchmarking. If you have specific questions please ask here, we're back from easter vacation now :-)

**Comment by kkraune** [comment]
http://docs.vespa.ai/documentation/performance/attribute-memory-usage.html is relevant then

**Comment by bratseth** [comment]
_Attributes_: In-memory forward structures. Just an array with one entry per document, where the entry is a pointer if variable length.
Attributes may optionally have _fast-search_ specified, in which case an in-memory B-tree is maintained over the data.

_Indexes_: Search-engine style reverse indexes, persisted on disk but partially or fully cached in memory (adaptively, depending on available memory). This is _dictionaries_, which are sorted lists of all existing tokens, pointing to _posting lists_, which have sorted entries for each occurrence of each token in each document. These are compressed. In some cases _bitvectors_ are used. For each index there is also one (or more) in-memory B-trees indexing recent changes (and _block lists_ to prevent older versions of documents showing up when terms are removed).

_Document store_: A blob store which stores the raw data of each document in compressed form. These are split into sorted chunks, somewhat similar to LevelDb.

Updates to indexes happens on in-memory B-trees plus a _transaction log_ for persistence. B-trees are flushed to persistent posting lists in the background, which are then merged in the background.

count(DISTINCT xxx) is supported, see http://docs.vespa.ai/documentation/reference/grouping-syntax.html#counting-unique-groups
That this uses a data sketch (HyperLogLog) such that we can compute a count across multiple content nodes (data shards) without exchanging the full list of unique values.

**Comment by sqsw** [comment]
@bratseth Thanks a lot for the detailed reply. For high-performance feed/query use case, we might want to store all the fields in memory, would you suggest using 'index' or 'attribute' or both on each field of search definition? It looks like 'index' (dictionary) is more efficient than attributes (by fast-search B-tree).  

Yup, I tried counting-unique-groups, it looks like a little slow (5s) over around 50M documents, but your pointer indicated this is a optimized solution.

**Comment by sqsw** [comment]
Thanks for the best practices. I ran the test over 4k groups with sufficient system resources.

**Comment by bratseth** [comment]
On the unique count performance: By default we only allocate one thread per content node per query. You could try to increase this to equal the number of hyperthreads on the node: http://docs.vespa.ai/documentation/content/setup-proton-tuning.html#requestthreads
Beyond that, you'll need to reduce the number of documents on the node (that is, add more content nodes) to increase performance.

**Comment by sqsw** [comment]
As far as performance is concerned, is the query latency decreased linearly with increasing the number of content nodes? could you share some data about this?

**Comment by sqsw** [comment]
As far as performance is concerned, is the query latency decreased linearly with increasing the number of content nodes? could you pls share some data about this?

**Comment by kkraune** [comment]
generally yes, but it's complicated :-)

More nodes means less documents to rank per search node, hence work per search node goes down and latency with it. But there are many things to consider, like making sure you use the nodes you have (threads) before you add more nodes. There are non-linearities (some features cuts evaluation early).

The best read is http://docs.vespa.ai/documentation/performance/sizing-search.html, let us know what you think

**Comment by sqsw** [comment]
@kkraune , thank you for the link. Did you happen to have some feeding (rps) performance results on a standalone node? Assume that the container and content node are deployed on a same host, all fields are indexed/stored in RAM (via attribute, or attribute | index), and the system resources(CPU/RAM) are sufficient to use.  

By the way, I was seeing huge usage of 'buff/cache' , was there any disk io even if I specified 'attribute' for all fields in sd? (redundancy=2)

**Comment by kkraune** [comment]
sorry we don't publish such numbers as they are only valid for a given data/query set on a given host. We don't have such a golden set as of now.

if you only hit attributes there should be no disk IO. you can use strace or similar tools to see the IO calls if that is relevant. Most often it is not if throughput and latency is good for the test

**Comment by bratseth** [comment]
@sqsw there will be disk IO even if you only write attributes because the data is persisted. See also http://docs.vespa.ai/documentation/reference/search-definitions-reference.html#fast-access

You should expect a few thousand to a couple tens of thousands of write operations per second, depending on hardware, network, data size etc.

**Comment by baldersheim** [comment]
All disk writes happen in the background and is append only so all disk access is sequential.
It is hard to give feed/indexing speed in documents/per second. It is a little easier to give in in MB/s.
Default setup will normally handle 10-20 MB/s of data to a field of type index or attribute with fast-search. Data that is summary only or goes to an attribute that is not 'fast-search' we see that 200+ MB/s is possible.
It is possible to tune indexing speed up by. Default tuning is conservative to prefer query traffic.
See below links for further reading.
http://docs.vespa.ai/documentation/writing-to-vespa.html
http://docs.vespa.ai/documentation/content/setup-proton-tuning.html#feeding-concurrency

**Comment by sqsw** [comment]
Thank you @bratseth, this explains the disk io activities in my env. Is there any option to avoid being persisted on disk? The searchable-copies equals to the redundancy (2), so I think all the replicas are in memory. Avoid writing disk is important for my high throughput scenario, do you have suggestion about that?

**Comment by kkraune** [comment]
you can use a ramdisk ;-)

**Comment by sqsw** [comment]
@baldersheim , could you pls elaborate more about why attribute with fast-search would result in lower throughput than summary only fields. Also, would the attribute(fast-search, B-tree) and indexes be written to disk as well, is the disk persistence (for attribute) designed for data recovery from reboot? For document auto-expiry, how to cleanup the data from a sequential persistence? Thanks!

**Comment by baldersheim** [comment]
1 - All data are written to a transaction log on arrival. This log is pruned when other internal structures are persisted in the background. On restart of the server this log is replayed to obtain same state as before going down either because of controlled restart, server crash, kernel panic, or powerloss.

2.1 - All document put operations are written to a document store. This is an append only store. Updates or removes will just write a new entry that automatically invalidates the old entry. The holes, also known as bloat, of unused/stale data are compacted away when bloat passes a threshold, default 20%.

2.2 Updates that does not contain attribute-only fields will incur a read from the document store and a write of a new version of the document.

3 - Attributes are fully kept in memory with regular persistence to disk. This persistence is to enable pruning the transaction log. Both the values themselves and the btrees are kept in memory, while during persistence only the values are persisted.

4 - Indexes store the index structures for reverse lookup on disk. This also happens in the background when the memory footprint is too high or it is worthwhile in order to keep a reasonable size of the transaction log.

5 - Except for the first store to the transaction log, all the other disk operations happen in the background and will take place after the ack has been sent that the operation was succesful.

6 - Building the btrees used by the attributes and the disk indexes in memory takes place before we send the ack. By default this happens in a single thread and is not batched to keep latency low. This operations happens on live structures and uses a Read-Modify-Write pattern to avoid locking or interfering with ongoing queries. These operations often has a high cpu cache-miss rate due to btrees and large dictionaries. This is normally what limits feed speed. But again, this is very data dependent.
This is what can be tuned by allowing batching with visibility-delay and indexing concurrency.

So in short, builing indexes/btrees happens inline with feeding, while writing document store happens in the background with batching all tricks to give high throughput.

Hope this helps.

**Comment by baldersheim** [comment]
1.1 - Yes, it will be duplicated and is not recommended. Then it will always search in the index, and use the attribute when doing grouping and other accesses that require the value itself. However fast-search is definitely waste of memory. I would probably recommend then duplicating the field into f_index with 'index' and then f_attribute with 'attribute'. That gives far better control. It is also advised to not use 'summary' on the field itself and instead use explicit summary classes.

1.2 Correct. Unless you specify visibility-delay. Then ack happens after it is written to transaction log. It will then be visible for search no later than the specified time in visibility delay. That allows batching.

1.3 'contains' might trigger a disk lookup, but will normally not. All disk indexes are memory mapped and frequently accessed terms will reside in memory after the first query pulls it into memory.

2. Only transaction log cause disk activity inline with feeding. However all background activity flushing documents, attributes, writing new index generations will cause disk access in the background.
If you run vespa-logfmt -l event | grep -i eventlogger will give you information on all background operations. It might be that the heuristics for when to run background operations are not optimal for your usecase. Number of cores, memory, disk throughput, and feed pattern are needed to make optimal tuning.
What feed speed do you get ?
What is the throughput on your IO system ?
How much memory do you have ?
How many cores does your machine have ?
If you attach the output from the command above we might be able to tell if any tuning might help.

**Comment by sqsw** [comment]
What feed speed do you get ?  ~21K rps, 150B per document
What is the throughput on your IO system ? SATA, around 100M/s
How much memory do you have ? 128G
How many cores does your machine have ? 42

I'd be interested to know if you have plans to:
1.  support 'join' across multiple document types;
2.  build (or already have) a query UI on top of vespa.

**Comment by bratseth** [comment]
1. There is support for that, see http://docs.vespa.ai/documentation/search-definitions.html#document-references

2. No plans. Do you mean for issuing queries as an aid for developers, or on top of the config servers as a tool for operators, or something else?

**Comment by sqsw** [comment]
@bratseth Thanks for the pointer, I'll take a look :)
I mean a frontend UI for people to choose/specify conditions (field range, groups, etc), then draw graph based on the query results, something like superset.

**Comment by sqsw** [comment]
For 1, I'd expect something like 'select count(*) from a, b where a.user_id=b.user_id and a.type=1 and b.status=0;'

**Comment by sqsw** [comment]
For 1, I'd expect something like 'select count(*) from a, b where a.user_id=b.user_id and a.type=1 and b.status=0;'

2. Yup, like an aid for dev to issue queries and generate graph based on query result.

**Comment by sqsw** [comment]
@bratseth , I looked into document references feature, it looks to me the main use case is to merge fields from existing document types. Does it support joining documents based on a few shared fields (i.e. user_id)?

**Comment by bratseth** [comment]
You need to know at configuration time what you'll join on and which fields you'll access from the joined document. Also, the document you join with need to have a global distribution. This is because Vespa is for use cases where you need low latency and scalability, and join in general is not possible to perform with low latency and is not scalable. If you are able to formulate your use case using document references be guarantee that it will perform and scale.

Another subset of join which is performant and scalable is joining only the top N documents (where N is reasonably low). To implement this, you can write a Searcher component which first retrieves the top N documents (as usual) and then performs a separate query (or use the Document API) to join in the referenced document.

Regarding the ui: I'm not sure what you mean by "generate graph", but if you want to create a UI on top of Vespa we are happy to take contributions.
